vars/deleteFile.groovy

```diff
@@ -0,0 +1,57 @@
+import com.sap.piper.ConfigurationHelper
```

Comment:
Does this step work in the free wild on your side?

On my side it does not work.

I have this pipeline script:

```
@Library('piper-lib-os') _

node() {
    
    stage('delete') {
        deleteDir()
        sh 'echo delete > deleteme.txt; pwd; ls -la; '
        deleteFile script: this, path: 'deleteme.txt'
        sh 'pwd; ls -la'
    }
}
```

Expected behaviour: After the delete step there is no file `deleteMe.txt` anymore
Acutual behaviour: After the delete step there is still that file (as we can see with the `ls`

I added this line of code in the method body of `deleteOutputFileIfPresent`:

```
echo "File: ${originalFile.getAbsolutePath()}"
```

That is resolved to: `File: /deleteme.txt`.

So I guess we have to take the current working directory into account. Correct me if I'm wrong

There is also no message that the file could not be deleted since we have a check if the file exists earlier. The file does not exist (inside `/`) and there is no attempt at all to delete the file.

By the way: In another thread we are discussion about how should be performed by mockes (should they really delete files). This is a good example for mocks giving a false feeling of security.

Reply:
... when thinking about more details here: how does this behave in case some code is dispatched to other nodes (might be running on Jenkins slaves). To my understanding the groovy code runs always on master. Is a File created on the master also suitable if the workspace is located on a slave? Not sure here ... Maybe we should test this. Other approach here: we can issue something like `sh 'rm $file`. I think the sh statement is executed on the slave (works only for relative pathes, of course) in case it turns out we need such a generic `deleteFile` step. Had a look into [this](https://github.com/jenkinsci/workflow-basic-steps-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/steps/DeleteDirStep.java): that is a similar step for recursivelmy deleting directories. They use some `getContext().get(FilePath.class).deleteRecursive();` (did not dive deeper ...)

Reply:
That's correct. Thanks for the hint! 
The `deleteFile` script required an absolute path which is not really nice.

This is fixed now, and no longer uses `java.io.File` but `sh rm`. Reading up on it, `java.io.File` can actually be dangerous since - as you said - all Groovy code is executed on the master and only the results are returned to remote nodes. Some people claim, `java.io.File` should only be used on Jenkins to assemble paths.

I also removed `java.io.File` usage from `cfManifestSubstituteVariables` and replaced it with `fileExists`. 

Also added missing tests for `deleteFile`. Your scenario from above now works in my setup.
Please retry and let me know what you think.

Reply:
I can confirm it is better now. There is one - easy to solve - issue remaining:
I would like to suggest to add single quotes
```
sh script: "rm '${filePath}' // mind the single quotes surrounding 'filePath'"
```

Otherwise a file name containing a space will be handled like two files _(You think that will not happen? I promise it will happen faster than you believe)_.

There is one other thing: I'm late to the party and I should have mentioned this earlier, but iI smelled blood before since the file delete failed caused by the `java.io.File` trouble.

I would like to suggest to put the `deleteFile` step into something like a `FileUtils` class located inside `src`. The steps exposed by the piper lib perform normally some heavier actions like deploying something, performing maven or mta builds etc. From that point of view a `deleteFile` step is very gereric. Hence I would like to suggest to move it into a helper class. Again: I apologize or raising this late. I should have done that earlier.

Maybe the same applies here: we have two replace-yaml steps now. The gereric one and that one for the cf yml. The cf yaml step forwards to the more generic one. Maybe we can also have the generic step not as step but as util class somewhere. With such an approach there is also no need to handle the return code via the common pipeline environment (that is needed since declarative pipelines does not support return values for steps; see `vars/cfManifestSubstituteVariables.groovy#106` in the current HEAD commit of this PR). What do you think?

The point behind is: the steps are API and intended to be used inside pipelines. When a step is out, we have to be backward-compatible. Anything located in `src` is not considered to be API and can be changed in an incompatible manner easily later.

Reply:
Ok, thanks. No need to convince me about the single quotes around the file path. Happens to me all the time... 😉 That's an easy fix and totally worthwhile.
I think I understand your other points, and I will give it a try tomorrow.
There is one thing I don't quite understand: what do you mean "declarative pipelines does not support return values for steps"? I searched for that a bit, and did not quite find anything about this. What's the reason behind that? And is this due to the architecture of this library or a general Jenkins advice?

Reply:
I thought about this a little more. And
1. I am not 100% sure what it would mean if I want to delete a file from within `steps{ }`.
    Could I still call to the `FileUtils` in `src` from the pipeline? 
1. There is a Jenkins basic step called `deleteDir` which kind of misses a counterpart to delete a simple file.  Wouldn't it be good to keep the `deleteFile` as a step then?

The second part about the `yamlSubstituteVariables` I think is more suited for being moved to a utils. Especially, since - as you explained - one cannot assign return values within pipelines (other than in `script` blocks) and hence it would be rather seldom to parse some yaml yourself and then have it injected into `yamlSubstituteVariables`.

Since also returning the result via the environment feels really awkward here, I suggest to move `yamlSubstituteVariables` to a utils class and keeping `deleteFile` as it is.

Reply:
Btw: where do you test the utils classes under `src`?

Reply:
in the packages inside `test/groovy`. The tests for the steps itself are contained in the default packge, the utils etc... comes with a package

Reply:
ad 1.) 

> Could I still call to the FileUtils in src from the pipeline?

in a pure declarative pipeline you cannot call to stuff from `src` (... inside a script block you can do this also in declarative pipelines). In scripted pipelines you can calls stuff from `src`. But by default we consider the coding contained in `src` not as API, this means it should not be used from pipelines (despite this is technically possible). At least so far there was no need to delete a file directly from a pipeline. Anyway we can call to coding contained in `src` from pipeline steps.

 ad 2.

> There is a Jenkins basic step called deleteDir which 

`deleteDir` is not the perfect couterpart. It always deletes the current directory, it is not possible to specifiy a (sub)directory which gets deleted (ok, we can wrap it into `dir { ./.}`).

> Wouldn't it be good to keep the deleteFile as a step then?

So far nobody needed it. The point is: we would like to delete the original file before writing the files with replacements via `writeYaml`.  What about having this inside the step? Earlier I suggested to provide the option to hand in already parsed yml (together with: " (or at least provide that as additional option)"). So when using this step directly from a pipeline I now agree this is simpler when working on files. I have another use case in mind where it would be interesting to re-use the step (updating an `mta.yml` in the `mtaBuild`step. Here it would be better for us to provide at least the parameters as already parsed yml). Maybe that can be resolved by a type check: if we receive a String we handle it as a file name, it if is a `Map` (or other types we expect as parsed yml)  we take it as already parsed yml.

A agree on having `yamlSubstituteVariables` as util. But I still think that `deleteFile` does not fit into the other steps of the piper lib (which are performing more heavy weight build operations (`mtaBuild`, deployment to various targets). On the other hand i'm not _strongly_ against it. More opinions from other colleagues in the piper area here would be appreciated.

Reply:
Thanks for the explanations!  
I pushed some changes which now turned `yamlSubstituteVariables` into a util class, and I think it is nicer that way. Thus you can now use it for whatever scenario you have where parsed Yaml is available. 

This change came with a few nicer simplifications but also some headaches - especially caused by this [very silly Jenkins behaviour](https://issues.jenkins-ci.org/browse/JENKINS-55113). But luckily I could resolve it in a simpler way.

I still think that having `deleteFile` as a step (not a util) is a good thing to have. First, I needed it (and I am not nobody ... 😄 ) and second, it makes it easier for newbies to delete files. Sure, they could just use the shell but why not make it a little more "readable". Therefore, I would like to leave `deleteFile` as a step.

>Earlier I suggested to provide the option to hand in already parsed yml (together with: " (or at least provide that as additional option)"). So when using this step directly from a pipeline I now agree this is simpler when working on files.

I admit you lost me a bit on that one, since I am not sure which step you are talking about.
If you were referring to changing the `cfManifestSubstituteVariables` step so that it can potentially deal with both files *and* parsed Yaml objects, then I would object (pun intended).
With `YamlUtils` available now, that allow you to input parsed Yaml, you can now also realize the `mta.yml` case - either by calling `YamlUtils` from a script-block or by calling it in code.  
I think the implementation of `cfManifestSubstituteVariables` is clean and focused on a particular use case now. I would not want to mix in more - rather reuse what now is in the utils.